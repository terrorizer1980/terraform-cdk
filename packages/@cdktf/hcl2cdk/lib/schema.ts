import * as z from "zod";

const outputConfig = z.array(
  z.object({
    value: z.string(),
    description: z.string().optional(),
    sensitive: z.boolean().optional(),
    depends_on: z.string().optional(),
  })
);
export type Output = z.infer<typeof outputConfig>;

const variableConfig = z.array(
  z.object({
    type: z.string().optional(),
    default: z.any().optional(),
    description: z.string().optional(),
    sensitive: z.boolean().optional(),
  })
);
export type Variable = z.infer<typeof variableConfig>;

export const schema = z.object({
  locals: z.array(z.any()).optional(),
  variable: z.record(variableConfig).optional(),
  output: z.record(outputConfig).optional(),
});
