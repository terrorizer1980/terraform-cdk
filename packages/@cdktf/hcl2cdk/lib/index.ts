import { parse } from "@cdktf/hcl2json";
import template from "@babel/template";
import generate from "@babel/generator";
import * as t from "@babel/types";
import prettier from "prettier";
import { schema, Output, Variable } from "./schema";

type ConvertOptions = {
  language: "typescript";
};

const optionalProperty = (key: string, item: undefined | any) =>
  `${item ? key + ': "' + item + '",' : ""}`;
const optionalProperties = (record: Record<string, any>) =>
  Object.entries(record).reduce(
    (carry, [key, item]) => carry + optionalProperty(key, item),
    ""
  );

function output(key: string, item: Output): t.Statement {
  const [{ value, description, sensitive }] = item;
  const out = template(`
  new TerraformOutput(this, "${key}", {
      value: "${value}",
      ${optionalProperties({
        description,
        sensitive,
      })}
  })
`)();

  return out as t.Statement;
}

function variable(key: string, item: Variable): t.Statement {
  // We don't handle type information right now
  const [{ type, ...props }] = item;

  return template(`
  new TerraformVariable(this, "${key}", {
      ${optionalProperties(props)}
  })
`)() as t.Statement;
}

export async function convert(
  filename: string,
  hcl: string,
  options: ConvertOptions
) {
  if (options.language !== "typescript") {
    throw new Error("Unsupported language used: " + options.language);
  }

  const json = await parse(filename, hcl);
  console.log(JSON.stringify(json));
  const plan = schema.parse(json);

  const variableAst = Object.entries(plan.variable || {}).map(([key, item]) =>
    variable(key, item)
  );
  const outputAst = Object.entries(plan.output || {}).map(([key, item]) =>
    output(key, item)
  );
  const { code } = generate(t.program([...variableAst, ...outputAst]) as any);
  return prettier.format(code, { parser: "babel" });
}
