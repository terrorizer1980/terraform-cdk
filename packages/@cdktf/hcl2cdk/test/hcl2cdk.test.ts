import { convert } from "../lib/index";

describe("convert", () => {
  describe("typescript", () => {
    it.each([
      {
        t: "output",
        hcl: `
        output "cidr_out" {
            value = "test"
        }`,
      },
      {
        t: "output-sensitive",
        hcl: `
        output "cidr_out" {
            value = "test"
            sensitive = true
        }`,
      },
      {
        t: "output-description",
        hcl: `
        output "cidr_out" {
            value = "test"
            sensitive = true
            description = "Best output"
        }`,
      },
      {
        t: "output-multiple",
        hcl: `
        output "first_cidr_out" {
            value = "first"
        }
        output "second_cidr_out" {
            value = "second"
        }`,
      },
      {
        t: "variable",
        hcl: `variable "weekday" {}`,
      },
      {
        t: "variable-default",
        hcl: `
        variable "availability_zone_names" {
            type    = list(string)
            default = ["us-west-1a"]
            description = "What AMI to use to create an instance"
        }`,
      },
    ])("handles $t configuration", async ({ hcl, t }) => {
      const code = await convert(`${t}.hcl`, hcl, { language: "typescript" });
      expect(code).toMatchSnapshot();
    });

    it.todo("handles variable configuration");
    it.todo("handles references");
    it.todo("handles provider configuration");
    it.todo("handles module configuration");
  });
});
