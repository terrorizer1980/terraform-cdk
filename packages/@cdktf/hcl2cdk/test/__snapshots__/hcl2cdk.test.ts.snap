// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert typescript handles $t configuration 1`] = `
"new TerraformOutput(this, \\"cidr_out\\", {
  value: \\"test\\",
});
"
`;

exports[`convert typescript handles $t configuration 2`] = `
"new TerraformOutput(this, \\"cidr_out\\", {
  value: \\"test\\",
  sensitive: \\"true\\",
});
"
`;

exports[`convert typescript handles $t configuration 3`] = `
"new TerraformOutput(this, \\"cidr_out\\", {
  value: \\"test\\",
  description: \\"Best output\\",
  sensitive: \\"true\\",
});
"
`;

exports[`convert typescript handles $t configuration 4`] = `
"new TerraformOutput(this, \\"first_cidr_out\\", {
  value: \\"first\\",
});
new TerraformOutput(this, \\"second_cidr_out\\", {
  value: \\"second\\",
});
"
`;

exports[`convert typescript handles $t configuration 5`] = `
"new TerraformVariable(this, \\"weekday\\", {});
"
`;

exports[`convert typescript handles $t configuration 6`] = `
"new TerraformVariable(this, \\"availability_zone_names\\", {
  default: \\"us-west-1a\\",
  description: \\"What AMI to use to create an instance\\",
});
"
`;
